File: tree/235. Lowest Common Ancestor of a Binary Search Tree/Solution.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

public class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode curr = root;

        int currVal, pVal = p.val, qVal = q.val;
        while (curr != null) {
            currVal = curr.val;

            if (pVal < currVal && qVal < currVal) {
                curr = curr.left;
            } else if (pVal > currVal && qVal > currVal) {
                curr = curr.right;
            } else {
                return curr;
            }
        }
        return null;
    }
}

File: tree/235. Lowest Common Ancestor of a Binary Search Tree/Solution.cpp


struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution
{
public:
    TreeNode *lowestCommonAncestor(TreeNode *root, TreeNode *p, TreeNode *q)
    {
        int val = root->val;
        int pVal = p->val;
        int qVal = q->val;
        if (pVal < val && qVal < val)
        {
            return lowestCommonAncestor(root->left, p, q);
        }
        else if (pVal > val && qVal > val)
        {
            return lowestCommonAncestor(root->right, p, q);
        }
        else
        {
            return root;
        }
    }
};